File: 2nd year/Laboratories/MAP/L9/CRUD/UI/Tui/StudentTui.cs
using System;
using System.Collections.Generic;
using CRUD.Domain;
using CRUD.Service;
using CRUD.Utils;

namespace CRUD.UI.Tui
{
    public class StudentTui : BaseTui
    {
        private readonly StudentService _service;

        public StudentTui(StudentService service)
        {
            _service = service;
            Initialize("Student TUI", new Dictionary<string, Action>
            {
                {"Create student", CreateStudent},
                {"Update student", UpdateStudent},
                {"Delete student", DeleteStudent},
                {"Read one student", ReadOneStudent},
                {"Read all students", ReadAllStudents},
                {"Mark student absent", MarkStudentAbsent},
                {"Justify student absence", JustifyStudentAbsence}
            });
        }

        private void CreateStudent()
        {
            var student = Read("Id", "Group", "Name", "Email", "TeacherName");
            Write(GenerateMessage(_service.CreateStudent(student)));
        }

        private void UpdateStudent()
        {
            var student = Read("Id", "Group", "Name", "Email", "TeacherName");
            Write(GenerateMessage(_service.UpdateStudent(student)));
        }

        private void DeleteStudent()
        {
            var id = Read("Id");
            Write(GenerateMessage(_service.DeleteStudent(id)));
        }

        private void ReadOneStudent()
        {
            var id = Read("Id");
            var student = _service.ReadOneStudent(id);
            Write(student == null
                ? "Student not found..."
                : EntityUtils.DumpString<Student, int>(student));
        }

        private void ReadAllStudents()
        {
            foreach (var student in _service.ReadAllStudents())
            {
                Write(EntityUtils.DumpString<Student, int>(student));
            }
        }

        private void MarkStudentAbsent()
        {
            var id = Read("Id");
            var week = Read("Week");
            Write(GenerateMessage(_service.MarkStudentAbsent(id, week)));
        }

        private void JustifyStudentAbsence()
        {
            var id = Read("Id");
            var week = Read("Week");
            Write(GenerateMessage(_service.JustifyStudentAbsence(id, week)));
        }
    }
}

File: 2nd year/Laboratories/MAP/7/src/main/java/ui/tui/StudentTUI.java
package ui.tui;

import service.StudentService;

import java.util.HashMap;
import java.util.Map;
import java.util.stream.Stream;

class StudentTUI extends BaseTUI {
    private StudentService studentService;

    StudentTUI(StudentService studentService) {
        this.studentService = studentService;
        generateTUI(
                "Student TUI",
                new HashMap<String, Runnable>() {{
                    put("Create student", StudentTUI.this::createStudent);
                    put("Read one student", StudentTUI.this::readOneStudent);
                    put("Read all students", StudentTUI.this::readAllStudents);
                    put("Update student", StudentTUI.this::updateStudent);
                    put("Delete student", StudentTUI.this::deleteStudent);
                    put("Mark student absent", StudentTUI.this::markStudentAbsent);
                    put("Justify student absence", StudentTUI.this::justifyStudentAbsence);
                }});
    }

    private void createStudent() {
        Map<String, String> studentMap = readMap("id", "group", "name", "email", "teacherName");
        System.console().printf("Operation %s\n", studentService.createStudent(studentMap) ? "successful" : "failed");
    }

    private void readOneStudent() {
        String studentId = readOne("id");
        Stream.of(studentService.readOneStudent(studentId))
                .filter(s -> !s.isEmpty())
                .forEach(s -> System.console().printf("%s\n", s));
    }

    private void readAllStudents() {
        studentService.readAllStudents()
                .forEach(s -> System.console().printf("%s\n", s));
    }

    private void updateStudent() {
        Map<String, String> studentMap = readMap("id", "group", "name", "email", "teacherName");
        System.console().printf("Operation %s\n", studentService.updateStudent(studentMap) ? "successful" : "failed");
    }

    private void deleteStudent() {
        String studentId = readOne("id");
        System.console().printf("Operation %s\n", studentService.deleteStudent(studentId) ? "successful" : "failed");
    }

    private void markStudentAbsent() {
        String studentId = readOne("studentId");
        String absenceWeek = readOne("absenceWeek");
        System.console().printf("Operation %s\n", studentService.markStudentAbsent(studentId, absenceWeek) ? "successful" : "failed");
    }

    private void justifyStudentAbsence() {
        String studentId = readOne("studentId");
        String absenceWeek = readOne("absenceWeek");
        System.console().printf("Operation %s\n", studentService.justifyStudentAbsence(studentId, absenceWeek) ? "successful" : "failed");
    }
}
